### PR TITLE
[opt](fqdn) Add DNS Cache for FE and BE

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

In previously, when enabling FQDN, Doris will call dns resolver to get IP from hostname
each time when 1) FE gets BE's grpc client. 2) BE gets other BE's brpc client.
So when in high concurrency case, the dns resolver be overloaded and failed to resolve hostname.

This PR mainly changes:

1. Add DNSCache for both FE and BE.
    The DNSCache will run on every FE and BE node. It has a cache, key is hostname and value is IP.
    Caller can get IP by hostname from this cache, and if hostname does not exist, it will try to resolve it
    and update the cache.
    In addition, DNSCache has a daemon thread to refresh the cache every 1 min, in case that the IP may
    be changed at anytime.

There are other implements of this dns cache:

1.  https://github.com/kaka11chen/doris/commit/36fed139974ed52dfa61b656f3e4d64f56a4185a
    This is for BE side, but it does not handle the IP change case.

3. https://github.com/apache/doris/pull/28479
    This is for FE side, but it can only work with Master FE. Other FE node will not be aware of the IP change.
    And there are a bunch of BackendServiceProxy, this PR only handle cache in one of them.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Review by kaka11chen — APPROVED** [review]
LGTM

File: be/src/util/dns_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include "util/dns_cache.h"
+
+#include "service/backend_options.h"
+#include "util/network_util.h"
+
+namespace doris {
+
+DNSCache::DNSCache() {
+    refresh_thread = std::thread(&DNSCache::_refresh_cache, this);
+    refresh_thread.detach();
```
Comment (zhiqiang-hhhh) [inline be/src/util/dns_cache.cpp:27]: the detached thread can not be join by main thread

https://stackoverflow.com/questions/39174761/pthreads-can-i-detach-from-a-thread-and-then-join-in-main

Comment (morningman) [reply be/src/util/dns_cache.cpp:27]: I will fix it

File: be/src/util/dns_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+DNSCache::~DNSCache() {
+    stop_refresh = true;
+    if (refresh_thread.joinable()) {
```
Comment (zhiqiang-hhhh) [inline be/src/util/dns_cache.cpp:32]: not joinable, so DNECache will be destoried immediately

File: be/src/util/dns_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+void DNSCache::_refresh_cache() {
+    while (!stop_refresh) {
+        // refresh every 1 min
+        std::this_thread::sleep_for(std::chrono::minutes(1));
+        std::unordered_set<std::string> keys;
+        {
+            std::shared_lock<std::shared_mutex> lock(mutex);
+            std::transform(cache.begin(), cache.end(), std::inserter(keys, keys.end()),
```
Comment (zhiqiang-hhhh) [inline be/src/util/dns_cache.cpp:74]: it is very likely that object cache has already been destroyed, if main thread has already exited with --grace enabled.

File: be/src/util/dns_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        cache[hostname] = real_ip;
+        LOG(INFO) << "update hostname " << hostname << "'s ip to " << real_ip;
+    }
+    return Status::OK();
+}
+
+void DNSCache::_refresh_cache() {
+    while (!stop_refresh) {
+        // refresh every 1 min
+        std::this_thread::sleep_for(std::chrono::minutes(1));
```
Comment (zhiqiang-hhhh) [inline be/src/util/dns_cache.cpp:70]: btw, after `_refresh_cache` thread is joinable, when we stop BE with grace enabled, we will wait at least 1 min ...
